[pre-ll] Support HLSL semantic indexes

tangmi opened the pull request:
Fixes #2748
PR checklist:
- [x] tested examples with the following backends: DX11

This changeset adds support for HLSL semantic indices.

Users should now be able to specify a semantic with an index (such as `"TEXCOORD1"`) in their pipeline. If one does not include the index, an index of `0` is assumed (as HLSL assumes).

Introduces a struct that can be converted to and from a packed String.

Here are some sample mappings:
- `"TEXCOORD"` -> `VertexSemantic { name: "TEXCOORD", index: 0 }` -> `"TEXCOORD0"`
- `"TEXCOORD0"` <-> `VertexSemantic { name: "TEXCOORD", index: 0 }`
- `"TEXCOORD10"` <-> `VertexSemantic { name: "TEXCOORD", index: 10 }`

On link_input, we will try linking against the semantic name and index together (e.g. `"TEXCOORD0"`) and also just the semantic name (e.g. `"TEXCOORD"`) if the semantic index is 0, in the case that the user did not explicitly provide the semantic index, esp. for GLSL compatibility).

kvark commented on src/backend/dx11/src/factory.rs:
```diff
@@ -589,7 +589,7 @@ impl core::Factory<R> for Factory {
             layouts.push(d3d11::D3D11_INPUT_ELEMENT_DESC {
                 SemanticName: &charbuf[charpos],
-                SemanticIndex: 0,
+                SemanticIndex: u32::from(attrib.slot),
```
why isn't this just `attrib.slot as u32` or something like this?

tangmi replied:
Clippy has yelled at me in the past about this: https://rust-lang.github.io/rust-clippy/master/#cast_lossless. Though I'm happy to change it to `as u32` if that's preferred.

msiglreith commented:
Just to be sure, this change is currently backwards incompatible and would require adjusting all the HLSL shaders, right?

tangmi commented:
I've done some work to this branch, as my previous attempt made it require a vertex attribute have a semantic index equal to the attribute slot assigned to it.

Ideally this change should be backwards compatible, but in the latest revision I *believe* I have fixed a bug in the DX backend regarding multiple vertex buffers in a single pipeline (though I don't have a test for this yet).

Here's a bit of a brain dump:

- `AttributeVar`'s `slot` doesn't seem to do anything on directx (I am unaware of being able to explicitly set registers on individual vertex attributes is possible?)
- `attrib.slot` was previously the value set for the `InputSlot`, but that's a field meant for the buffer binding index when there are multiple vertex buffers (https://docs.microsoft.com/en-us/windows/desktop/api/d3d11/ns-d3d11-d3d11_input_element_desc)
  - I'm unsure how this worked in the past
- My current solution involves adding a "semantic_index" field on `AttributeVar` (that is set to 0 on all other platforms)
  - I tried to reuse `attrib.slot` for the semantic index, but it looks like that value is used as the index for some arrays in the DX impl (and setting it to existing values on multiple attributes causes indices in those arrays to be overridden)
- I don't like that my solution for `link_input` is "try the semantic, then try the semantic with the index, then fail"

tangmi commented on src/backend/dx11/src/factory.rs:
```diff
@@ -630,7 +630,33 @@ impl core::Factory<R> for Factory {
         };
         if !winerror::SUCCEEDED(hr) {
-            error!("Failed to create input layout from ?, error {:x}"/*, layouts*/, hr);
+            struct DebugInputLayout<'a> {
```
I can remove this if it's unsightly.

I'm working on https://github.com/retep998/winapi-rs/pull/760, which would additionally make this unnecessary.

kvark commented:
@tangmi thank you for the investigation! It would take me a bit of time to remember the internals of pre-ll DX11 backend. Overall, don't put too much trust it in - it was likely rolled out with the examples being the only test case. Saying that, the instancing example has multiple vertex buffers, which according to your findings shouldn't be working correctly, but it does.

tangmi commented:
Unfortunately I think the instancing example just happens to work. If I reorder the vertex attributes in the shader, the example stops displaying output.

Here are some observations:
- It looks like DX's input element's input slot is clamped to the range of bound vertex buffers
- Gfx reads the AttributeVar::slot from the order of the HLSL reflection data

So if the following vertex buffers are bound in the instance example
```
vertex buffer 0 => position data
vertex buffer 1 => translate and color instance data
```

The HLSL has a vertex defined as
```
struct VsInput {
	float2 pos: a_Position; // <-- will be assigned AttributeVar::slot = 0
	float2 trans: a_Translate; // <-- will be assigned AttributeVar::slot = 1
	uint color: a_Color; // <-- will be assigned AttributeVar::slot = 2
};
```

Then:
- `a_Position`'s InputSlot will get set to 0 (vertex data)
- `a_Translate`'s InputSlot will get set to 1 (instance data)
- `a_Color`'s InputSlot will get set to 2, but clamped to 1 (instance data)

And the example works

Reordering the vertex attributes in the shader to something like:
```
struct VsInput {
	float2 trans: a_Translate; // <-- will be assigned AttributeVar::slot = 0
	float2 pos: a_Position; // <-- will be assigned AttributeVar::slot = 1
	uint color: a_Color; // <-- will be assigned AttributeVar::slot = 2
};
```

Then:
- `a_Translate`'s InputSlot will get set to 0 (vertex data) <-- incorrectly set
- `a_Position`'s InputSlot will get set to 1 (instance data) <-- incorrectly set
- `a_Color`'s InputSlot will get set to 2, but clamped to 1 (instance data)

tangmi commented:
@kvark I think this is ready to review. All the examples run without modification (excluding `render_target`, which is missing HLSL shaders).

My reservations with these changes are:
- Adding a DX-only `semantic_index` field to `AttributeVar` is hacky
- `link_input` could be implemented better (i.e. `T::query` could generate code to look for semantic indices instead of `link_input` trying both, but idk how well this would work for code that manually implements `query`)

kvark commented:
Thank you for evolving the PR into shippable state! Could you please rebase instead of introducing the merge commit?
Also, does @msiglreith want to have a look before we proceed?

msiglreith commented:
So, the semantic name can basically be chosen randomly, right? In case of SPIRV-Cross outputted shaders, user need to be aware of the ordering of the vertex input. They need to match the declaration on the host side. Does my understanding match with the code?

tangmi commented:
The semantic name still needs to be set correctly, unfortunately. Since HLSL can't bind attributes by index, SPIRV-Cross maps the binding location to index of a consistent, but arbitrary, semantic: `TEXCOORD`.

As an aside, I believe any semantic could've been chosen, but I think people were mostly already using `TEXCOORD` for packing custom data in DX9/10. In DX11+, I think any semantic that doesn't conflict with the `SV_` ones would work.

tangmi commented:
@kvark This should be rebased and ready for re-review! Please let me know if there's anything else I should do.

kvark commented on src/core/src/shade.rs:
```diff
@@ -360,6 +360,8 @@ impl From<Stage> for Usage {
     /// Name of this attribute.
     pub name: String,
+    /// Hack: Semantic index used by DirectX. Should be set to 0 in all other backends.
+    pub semantic_index: u32,
```
this is a bit unfortunate - would require a new minor release for the whole project. Is there a way for us to only limit the changes to dx11 backend?

kvark replied:
ping @tangmi

tangmi replied:
Hi! Sorry for the late reply!

The only thing I can think of is making `AttributeVar` parameterized over a "name" type that dx11 could specialize as `(String, u32)` and other backends would use `String`. This would only ensure that the size/layout of the data doesn't change for the other backends. I'm not sure if this is sufficient to avoid a minor release for whole project?

`DataLink::link_input` for `VertexBufferCommon` needs the semantic index, but it only gets `AttributeVar`. `Factory::create_pipeline_state_raw` also needs the semantic index, and `AttributeVar` seems like the only thing that makes sense to hold it.

kvark replied:
Sorry, I'm still trying to figure out what the best solution here is. So here is another question: would it be feasible for your case to require the user to *only* use semantic index == 0 in their shaders? It seems to me that it would be compatible with the current code. What would be the downside of this path?

tangmi replied:
I'm not sure I understand the question fully, sorry!

Here's an alternate implementation that might be better: https://github.com/tangmi/gfx/pull/5 It doesn't change `AttributeVar` at the cost of very slightly higher cost marshaling in/out of that struct for the DX11 backend and slightly more complex vertex attribute handling on the DX11 backend.

tangmi commented:
@kvark I've rebased the new changes into this branch and updated the PR comment at the top. There *should* be no actual changes from tangmi#5, but please sanity check me if you think it's a good idea!
